packages/stream-model-instance-handler/src/model-instance-document-handler.ts

```diff
@@ -308,4 +317,37 @@ export class ModelInstanceDocumentHandler implements StreamHandler<ModelInstance
       }
     }
   }
+
+  /**
+   * Validates the ModelInstanceDocument unique constraints against the Model definition.
+   * @param model - model that this ModelInstanceDocument belongs to
+   * @param metadata - ModelInstanceDocument metadata to validate
+   * @param content - ModelInstanceDocument content to validate
+   */
+  async _validateUnique(
+    model: Model,
+    metadata: ModelInstanceDocumentMetadata,
+    content: Record<string, unknown> | null
+  ): Promise<void> {
+    if (model.content.accountRelation.type === 'set') {
+      if (metadata.unique == null) {
+        throw new Error('Missing unique metadata value')
+      }
+      if (content == null) {
+        throw new Error('Missing content')
+      }
+
+      const unique = model.content.accountRelation.fields
+        .map((field) => {
+          const value = content[field]
+          return value ? value.toString() : ''
+        })
+        .join('|')
+      if (unique !== toString(metadata.unique)) {
```

Comment:
Yes that's why all the values are cast to string 3 lines above, and the `unique` metadata value is an array of strings.

I don't think it's a problem that `1` (integer) is converted to `"1"` (string) as the fields types are enforced by the JSON schema anyway, as long as the conversion to string is stable. Maybe we can have an additional check in the model creation to ensure the SET fields are only defined on scalar fields?

Reply:
I think if instead of the `unique` field in the metadata always being a single string, it was allowed to be an array where each entry corresponded to one of the unique fields, then we could preserve the type information of each field and also allow things like StreamIDs to be used as unique fields

Reply:
The `unique` field in the metadata is a `Uint8Array`, here the logic converts all values to UTF-8 strings and then convert the concatenated string to `Uint8Array` to ensure the encoding is deterministic, but the purpose of the `unique` field is only to enforce determinism, no? How would it be needed or useful to have any type information about the values there?

Reply:
Well I'm thinking from the perspective of enforcing that the unique fields can't be changed.  Those fields should be immutable (essentially there's an implicit `@immutable` tag on them like we discussed in the field locking spec).  If the user can change the value from the number `1` to the string `"1"`, now that field isn't guaranteed immutable anymore.

That actually makes me think, do we actually have checks that these aren't changed?  We check that they match the corresponding values in the content, but what if there's a commit that updates the metadata and the content at the same time?  We should make sure we have a test for that and I don't believe the code will catch that case currently, unless I missed it.

Reply:
Immutability is more a side-effect of the deterministic behavior, the validation here is for the `unique` data, not so much the field types or other constraints.
Considering the work on immutable fields happens in parallel, what do you think if we make it part of the immutable fields logic to implicitly enforce the validation for the SET account relation fields?

Reply:
Um, I do feel like the immutability is pretty core to the feature.  If you can change the value of the field you are using for the set semantics, the correct behavior of the system gets pretty hard to reason about.  I'm fine with doing a follow-up PR to add enforcement of that immutability, but I wouldn't really want to release SET account relation without it.

Reply:
Actually, looks like we are already protected from this: https://github.com/ceramicnetwork/js-ceramic/blob/fa3e28c63b268359af5349a58dbdfea1d8992c53/packages/stream-model-instance-handler/src/model-instance-document-handler.ts#L155-L161.  So no concern here after all then.

Reply:
You mean mutability of the `unique` field? If so yes for sure it should be immutable, I thought you were talking about the fields from which the `unique` field is derived that would need stronger validation than checking their string values, should be all good then 👍

Reply:
oh you're right, I did mean that 😅.  I forgot what my concern was 😛 

There is still the case where if the json schema allows string or integer, the user could change the value from one to the other without any error.  That said, this approach we're using here would already make it so that if the user had a set account relation on such a field, the string value MID and number value MID would resolve to the same stream because the unique metadata for them would look identical.  So that seems like actually the bigger problem and extra validation on commit application wouldn't help with that.  I'm still not thrilled about it, and would prefer we use a metadata encoding with stronger typing.

But also, type safety in javascript is kind of a joke already anyway.  I think in the future we're going to need streamtypes that aren't based on jsonPatch, to not use json-to-string serialization/deserialization in the http api, and use something with much stronger type guarantees (like Protocol Buffers) as the underlying data encodings throughout.

As long as we remain in the javascript-centric world we are in right now, I think few of our users are likely to have real problems as a result of this.  So I think I'm fine moving forward with this as-is for now, even though it makes me a bit sad.